src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/LinkedEditingRange/LinkedEditingRangeEndpoint.cs

```diff
@@ -0,0 +1,153 @@
+            CancellationToken cancellationToken)
+        {
+            var document = await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(() =>
+            {
+                if (!_documentResolver.TryResolveDocument(
+                    request.TextDocument.Uri.GetAbsoluteOrUNCPath(), out var documentSnapshot))
+                {
+                    return null;
+                }
+
+                return documentSnapshot;
+            }, cancellationToken);
+
+            if (document is null || cancellationToken.IsCancellationRequested)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var location = await GetSourceLocation(request, document).ConfigureAwait(false);
+
+            // We only care if the user is within a TagHelper or HTML tag with a valid start and end tag.
+            if (TryGetMarkupNameTokens(codeDocument, location, out var startTagNameToken, out var endTagNameToken) &&
+                (startTagNameToken.Span.Contains(location.AbsoluteIndex) || endTagNameToken.Span.Contains(location.AbsoluteIndex)))
+            {
+                var startSpan = startTagNameToken.GetLinePositionSpan(codeDocument.Source);
+                var endSpan = endTagNameToken.GetLinePositionSpan(codeDocument.Source);
+                var ranges = new Range[2] { startSpan.ToRange(), endSpan.ToRange() };
+
+                return new LinkedEditingRanges
+                {
+                    Ranges = ranges,
+                    WordPattern = WordPattern
+                };
+            }
+
+            return null;
+
+            static async Task<SourceLocation> GetSourceLocation(
+                LinkedEditingRangeParams request,
+                DocumentSnapshot document)
+            {
+                var sourceText = await document.GetTextAsync().ConfigureAwait(false);
+                var linePosition = new LinePosition(request.Position.Line, request.Position.Character);
+                var hostDocumentIndex = sourceText.Lines.GetPosition(linePosition);
+                var location = new SourceLocation(hostDocumentIndex, request.Position.Line, request.Position.Character);
+
+                return location;
+            }
+
+            static bool TryGetMarkupNameTokens(
+                RazorCodeDocument codeDocument,
+                SourceLocation location,
+                [NotNullWhen(true)] out SyntaxToken? startTagNameToken,
+                [NotNullWhen(true)] out SyntaxToken? endTagNameToken)
+            {
+                var syntaxTree = codeDocument.GetSyntaxTree();
+                var change = new SourceChange(location.AbsoluteIndex, length: 0, newText: "");
+                var owner = syntaxTree.Root.LocateOwner(change);
+                var element = owner.FirstAncestorOrSelf<MarkupSyntaxNode>(
```

Comment:
Out of curiosity does WebTools implementation also return the surrounding start/end tag ranges?

Aka if you're here the start/end tag ranges will be returned:
```HTML
<div>
    |
</div>
```